add blog post for pulumi state move (#12286)
pulumi/docs: content/blog/move-resources-between-stacks/index.md

}, {
    parent: b,
});

const randomSite = new aws.s3.BucketObject("random.html", {
    bucket: b.bucket,
    content: randomPet.id,
}, {
    parent: b,
});
```

We create a bucket, and two bucket objects, both of them having the bucket as their parent.  One of the objects has a random pet name as its content.  At some point we decide we'd rather group all the AWS resources in a separate program, because this one grew too large.  We can do that using the `pulumi state move` command, and then adjusting the code.

First we can find out which stacks already exist and we can move the resources to:

```shell
$ pulumi stack ls --all
NAME                                            LAST UPDATE     RESOURCE COUNT  URL
v-thomas-pulumi-corp/pulumi-demo-move-aws/dev   23 minutes ago  0               <url>
v-thomas-pulumi-corp/pulumi-demo-move/dev       6 minutes ago   7               <url>
```

Next we can actually move the resources.  Since we're moving the resources from the currently selected stack, we can omit the `--source` argument:

```
$ pulumi state move --dest v-thomas-pulumi-corp/pulumi-demo-move-aws/dev 'urn:pulumi:dev::pulumi-demo-move::aws:s3/bucket:Bucket::b'

> Planning to move the following resources from dev to dev:

Author: Need to use the full stack name here.  Also gonna fix this.

  - urn:pulumi:dev::pulumi-demo-move::aws:s3/bucket:Bucket::b
  - urn:pulumi:dev::pulumi-demo-move::aws:s3/bucket:Bucket$aws:s3/bucketObject:BucketObject::index.html
  - urn:pulumi:dev::pulumi-demo-move::aws:s3/bucket:Bucket$aws:s3/bucketObject:BucketObject::random.html

The following resources being moved to dev have dependencies on resources in dev:

>   urn:pulumi:dev::pulumi-demo-move::random:index/randomPet:RandomPet::a-random-pet has a dependency on urn:pulumi:dev::pulumi-demo-move::random:index/randomPet:RandomPet::a-random-pet

Author: I hadn't noticed before, but this doesn't quite look right.  Gonna look into fixing this.

# `id` properties of the team member files at /data/team/team. Create a file for
# yourself if you don't already have one.
authors:
    - thomas-gummerer

# At least one tag is required. Lowercase, hyphen-delimited is recommended.
tags:
    - announcement
    - CLI

# The social copy used to promote this post on Twitter and Linkedin. These
# properties do not actually create the post and have no effect on the
# generated blog page. They are here strictly for reference.

# Here are some examples of posts we have made in the past for inspiration:
# https://www.linkedin.com/feed/update/urn:li:activity:7171191945841561601
# https://www.linkedin.com/feed/update/urn:li:activity:7169021002394296320
# https://www.linkedin.com/feed/update/urn:li:activity:7155606616455737345
# https://twitter.com/PulumiCorp/status/1763265391042654623
# https://twitter.com/PulumiCorp/status/1762900472489185492
# https://twitter.com/PulumiCorp/status/1755637618631405655

social:
    twitter:
    linkedin:

# See the blogging docs at https://github.com/pulumi/docs/blob/master/BLOGGING.md
# for details, and please remove these comments before submitting for review.
---

> Today we're announcing the [`pulumi state move`](https://www.pulumi.com/docs/cli/commands/pulumi_state_move/) command, which can be used to move resources that are managed by Pulumi between different stacks and/or projects.  No matter in which stack/project the resource was created in, it can now be moved with a single commmand.

Reviewer: Nit: These links don't need to be fully qualified:

```suggestion
Today we're announcing the [`pulumi state move`](/docs/cli/commands/pulumi_state_move/) command, which can be used to move resources that are managed by Pulumi between different stacks and/or projects.  No matter in which stack/project the resource was created in, it can now be moved with a single commmand.
```

<!--more-->

> While moving resources between projects is currently possible with some major surgery (our users came up with [automated solutions](https://github.com/pulumi/pulumi/issues/3389#issuecomment-679020482)), it's fairly error prone and difficult.  This makes refactoring Pulumi programs quite cumbersome, and having a simpler, integrated solution for this problem was a [much requested feature](https://github.com/pulumi/pulumi/issues/3389).

Reviewer 2: We could add some motivation here why you want to move resources:

```suggestion
When refactoring a Pulumi project from a [monolithic](https://www.pulumi.com/docs/using-pulumi/organizing-projects-stacks/#monolithic) structure to [micro stacks](https://www.pulumi.com/docs/using-pulumi/organizing-projects-stacks/#micro-stacks), you might need to move resources between different projects or stacks, without recreating them. While this has been possible with some major surgery (our users came up with [automated solutions](https://github.com/pulumi/pulumi/issues/3389#issuecomment-679020482)), it's fairly error prone and difficult.  This makes refactoring Pulumi programs quite cumbersome, and having a simpler, integrated solution for this problem was a [much requested feature](https://github.com/pulumi/pulumi/issues/3389).
```

Author: I like it, thanks!

splitting a stack into multiple stacks or when merging multiple stacks into one.

Usage:
  pulumi state move [flags] <urn>...

Flags:
      --dest string       The name of the stack to move resources to
  -h, --help              help for move
      --include-parents   Include all the parents of the moved resources as well
      --source string     The name of the stack to move resources from
  -y, --yes               Automatically approve and perform the move
```

Both `dest` and `source` can be either stacks in the current project, or stacks in a different project, using the fully qualified stack names.  Note that this works only for stacks within the same backend, it is currently not possible to move a resource between different backends.

The resources being moved have to be specified by their full URN, and multiple URNs can be passed at once.  For each resource being moved, all the children of that resource will also be moved, and the relationships between all resources being moved is preserved.  Resources with other types of dependencies will however not be moved to the target stack by default.

> The command will make no attempt to adjust the users program.  Inputs and outputs from broken dependencies will have to be adjusted in the program where the resource is being managed after being moved.  This can be done either by using stack references, or recreating the inputs in the program.

Reviewer 2: ```suggestion
The command will make no attempt to adjust the users program.  Inputs and outputs from broken dependencies will have to be adjusted in the program where the resource is being managed after being moved.  This can be done either by using [stack references](https://www.pulumi.com/docs/concepts/stack/#stackreferences), or recreating the inputs in the program.
```

> The resources being moved have to be specified by their full URN, and multiple URNs can be passed at once.  For each resource being moved, all the children of that resource will also be moved, and the relationships between all resources being moved is preserved.  Resources with other types of dependencies will however not be moved to the target stack by default.

Reviewer: ```suggestion
The resources being moved have to be specified by their full [URN](/docs/concepts/resources/names/#urns), and multiple URNs can be passed at once.  For each resource being moved, all the children of that resource will also be moved, and the relationships between all resources being moved is preserved.  Resources with other types of dependencies will however not be moved to the target stack by default.
```

> The resources being moved have to be specified by their full URN, and multiple URNs can be passed at once.  For each resource being moved, all the children of that resource will also be moved, and the relationships between all resources being moved is preserved.  Resources with other types of dependencies will however not be moved to the target stack by default.

Reviewer: It might be worth adding a sentence or two on how to find the URN of existing resources, e.g. via `pulumi stack --show-urns`.

Author: Done :heavy_check_mark:

> The resources being moved have to be specified by their full URN, and multiple URNs can be passed at once.  For each resource being moved, all the children of that resource will also be moved, and the relationships between all resources being moved is preserved.  Resources with other types of dependencies will however not be moved to the target stack by default.

Reviewer: Not sure if you want to get into it, but it might be worth mentioning that you may need to surround the URN with single quotes, as some URNs contain `$`, which can be misinterpreted by the shell.

Author: Yes, I think this is worth mentioning. Done :heavy_check_mark:

> The `pulumi state move` command is available in Pulumi v3.124.0 and later.

Reviewer: Should we update this now that there's a more recent release with additional fixes?

```suggestion
The `pulumi state move` command is available in Pulumi v3.126.0 and later.
```

Author: Yup, good idea!
